src/main/java/us/abstracta/wiresham/VirtualTcpService.java

```diff
@@ -124,24 +155,72 @@ private synchronized void removeClient(ConnectionFlowDriver connectionDriver) {
     connectionDrivers.remove(connectionDriver);
   }
 
-  public void stop(long timeoutMillis) throws IOException, InterruptedException {
+  private void handleSocketIOException(IOException e) {
+    if (stopped) {
+      LOG.trace("Received expected exception when server socket has been closed", e);
+    } else {
+      LOG.error("Problem waiting for client connection. Keep waiting.", e);
+    }
+  }
+
+  public void stop(long timeoutMillis) throws InterruptedException {
     synchronized (this) {
       stopped = true;
-      server.close();
       connectionDrivers.forEach(c -> {
         try {
-          c.close();
+          c.closeFlowConnections();
         } catch (IOException e) {
-          LOG.error("Problem closing connection {}", c.getId(), e);
+          LOG.error("Problem closing connection ", e);
         }
       });
     }
     clientExecutorService.shutdown();
-    clientExecutorService.awaitTermination(timeoutMillis, TimeUnit.MILLISECONDS);
-    serverExecutorService.shutdownNow();
-    if (!serverExecutorService.awaitTermination(timeoutMillis, TimeUnit.MILLISECONDS)) {
-      LOG.warn("Server thread didn't stop after {} millis", timeoutMillis);
+    if (!clientExecutorService.awaitTermination(timeoutMillis, TimeUnit.MILLISECONDS)) {
+      clientExecutorService.shutdownNow();
     }
   }
 
+  private FlowConnectionProvider buildFlowConnectionProvider() {
+    return new FlowConnectionProvider() {
+      public final Map<Integer, CompletableFuture<FlowConnection>> map = new ConcurrentHashMap<>();
+
+      @Override
+      public FlowConnection get(int port) throws ExecutionException, InterruptedException {
+        return map.get(port).get();
+      }
+
+      @Override
+      public void init(List<Integer> ports, FlowConnection flowConnection) {
+        map.clear();
+        for (Integer port : ports) {
+          map.put(port, new CompletableFuture<>());
+        }
+        CompletableFuture<FlowConnection> completedFuture = new CompletableFuture<>();
+        completedFuture.complete(flowConnection);
+        map.put(flowConnection.getPort(), completedFuture);
+      }
+
+      @Override
+      public void assignFlowConnection(int port, FlowConnection flowConnection) {
+        map.get(port).complete(flowConnection);
+      }
+
+      @Override
+      public boolean requiresFlowConnection(int port) {
+        return !map.get(port).isDone();
+      }
+
+      @Override
+      public void closeConnections() throws IOException {
+        for (CompletableFuture<FlowConnection> value : map.values()) {
```

Comment:
Don't you need to cancle the futures before closing them? Otherwise wouldn't it get stuck waiting for connections to sockets before closing them?

Reply:
Additionally, don't you need to synchronize access or something? I haven't though about it, but have you considered what happens when a connection is being established when the connections are being closed?